### PR TITLE
Implemented Item Properties View

### LINKED ISSUE #62: Add Item Properties view
The `theia-trace-extension` currently has a separate view in the explorer (left side) that populates upon selection in the graphs. For example, selecting an event in the `Events Table` the details will listed there. This should be supported in this project as well. An additional web-view needs to be implemented for that. Note that the `traceviewer-react-components` needs to provide the view implementation so that the web-view only need so use it and provide UI controls.

image

### PR DESCRIPTION
**Pull request by ngondalia**
Changes made:

- Uses an theia independent react widget for Properties view from traceviewer-react-components
- Proper signal handling between item properties webview and extension

This will add the Item Properties View to the Trace Viewer.

Fixes #62

Signed-off-by: Neel Gondalia ngondalia@blackberry.com

### CODE REVIEW

**Comment by ngondalia** [comment]
@bhufmann Please review

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        };
+        this._signalHandler = new VsCodeMessageManager();
+
+        window.addEventListener('message', event => {
+            const message = event.data; // The JSON data our extension sent
+            switch (message.command) {
+            case 'receivedTooltip':
+                console.log('got message');
+                const data = message.data;
+                console.log(data);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:34]: Remove console.log

File: vscode-trace-extension/src/trace-explorer/views/trace-explorer-properties-view-webview-provider.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/***************************************************************************************
```
Comment (bhufmann) [inline vscode-trace-extension/src/trace-explorer/views/trace-explorer-properties-view-webview-provider.ts:1]: This comment is for the `properties` directory. This directory should be under `trace-explorer` directly. The subdirectory `views` is the directory for the `available views` webview. I should have named it differently, maybe `available-views` to be clearer. Sorry, for the confusion. 

Please move `properties` and content to one level up under `trace-explorer` directory.

File: vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts [hunk 3]
```diff
@@ -134,6 +134,8 @@ export class TraceViewerPanel {
 	                this.loadTheme();
 	            }
 	            return;
+	        case 'updateTooltip':
```
Comment (bhufmann) [inline vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts:137]: update to `updateProperties`

File: vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts [hunk 4]
```diff
@@ -134,6 +134,8 @@ export class TraceViewerPanel {
...
+	            vscode.commands.executeCommand('messages.post.propertiespanel', 'receivedTooltip', message.data);
```
Comment (bhufmann) [inline vscode-trace-extension/src/trace-viewer-panel/trace-viewer-webview-panel.ts:138]: update to `receivedProperties`

File: vscode-trace-webviews/src/common/vscode-message-manager.ts [hunk 5]
```diff
@@ -71,4 +71,8 @@ export class VsCodeMessageManager extends Messages.MessageManager {
         const descWrapper = JSONBig.stringify(payload.getOutputDescriptor());
         vscode.postMessage({command: 'outputAdded', data: {data: expWrapper, descriptor: descWrapper }});
     }
+
+    tooltipUpdated(tooltip: { [key: string]: string }): void {
```
Comment (bhufmann) [inline vscode-trace-webviews/src/common/vscode-message-manager.ts:75]: Rename tooltip to properties

File: vscode-trace-webviews/src/common/vscode-message-manager.ts [hunk 6]
```diff
@@ -71,4 +71,8 @@ export class VsCodeMessageManager extends Messages.MessageManager {
...
+        vscode.postMessage({command: 'updateTooltip', data: {tooltip}});
```
Comment (bhufmann) [inline vscode-trace-webviews/src/common/vscode-message-manager.ts:76]: update tooltip to properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/index.tsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
+/***************************************************************************************
+ * Copyright (c) 2023 BlackBerry Limited and contributors.
+ *
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ ***************************************************************************************/
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import './index.css';
+import TraceExplorerProperties from './vscode-trace-explorer-tooltip-widget';
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/index.tsx:9]: update tooltip to properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+ * Licensed under the MIT license. See LICENSE file in the project root for details.
+ ***************************************************************************************/
+/* eslint-disable @typescript-eslint/ban-types */
+import React from 'react';
+import { VsCodeMessageManager } from 'vscode-trace-webviews/src/common/vscode-message-manager';
+import { ReactItemPropertiesWidget } from 'traceviewer-react-components/lib/trace-explorer/trace-explorer-properties-widget';
+import { signalManager } from 'traceviewer-base/lib/signals/signal-manager';
+
+interface PropertiesViewState {
+    tooltip: { [key: string]: string };
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:13]: properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+class TraceExplorerProperties extends React.Component<{}, PropertiesViewState> {
+    static ID = 'trace-explorer-tooltip-widget';
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:17]: properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    static LABEL = 'Item Properties';
+    private _signalHandler: VsCodeMessageManager;
+
+    constructor(props: {}) {
+        super(props);
+        this.state = {
+            tooltip: {}
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:24]: properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        };
+        this._signalHandler = new VsCodeMessageManager();
+
+        window.addEventListener('message', event => {
+            const message = event.data; // The JSON data our extension sent
+            switch (message.command) {
+            case 'receivedTooltip':
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:31]: Properties

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                console.log('got message');
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:32]: remove console.log

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                const data = message.data;
+                console.log(data);
+                signalManager().fireItemPropertiesSignalUpdated(message.data.tooltip);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:35]: properties

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx [hunk 14]
```diff
@@ -34,6 +34,12 @@ class App extends React.Component<{}, VscodeAppState>  {
 
   private _signalHandler: VsCodeMessageManager;
 
+  private _onTooltip = (tooltip: { [key: string]: string }): void => this.doHandleTooltipSignal(tooltip);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:37]: rename to use properties in the names

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx [hunk 15]
```diff
@@ -34,6 +34,12 @@ class App extends React.Component<{}, VscodeAppState>  {
...
+  /** Signal Handlers */
+  private doHandleTooltipSignal(tooltip: { [key: string]: string }) {
+      this._signalHandler.tooltipUpdated(tooltip);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:40]: properties

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx [hunk 16]
```diff
@@ -34,6 +34,12 @@ class App extends React.Component<{}, VscodeAppState>  {
 
   private _signalHandler: VsCodeMessageManager;
 
+  private _onTooltip = (tooltip: { [key: string]: string }): void => this.doHandleTooltipSignal(tooltip);
+  /** Signal Handlers */
+  private doHandleTooltipSignal(tooltip: { [key: string]: string }) {
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:39]: properties

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx [hunk 17]
```diff
@@ -74,16 +80,19 @@ class App extends React.Component<{}, VscodeAppState>  {
...
       });
+
       this.onOutputRemoved = this.onOutputRemoved.bind(this);
       this.onOverviewRemoved = this.onOverviewRemoved.bind(this);
       signalManager().on(Signals.OVERVIEW_OUTPUT_SELECTED, this._onOverviewSelected);
   }
 
   componentDidMount(): void {
       this._signalHandler.notifyReady();
+      signalManager().on(Signals.ITEM_PROPERTIES_UPDATED, this._onTooltip);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:91]: properties

File: vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx [hunk 18]
```diff
@@ -74,16 +80,19 @@ class App extends React.Component<{}, VscodeAppState>  {
...
   }
 
   componentWillUnmount(): void {
+      signalManager().off(Signals.ITEM_PROPERTIES_UPDATED, this._onTooltip);
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-viewer/vscode-trace-viewer-container.tsx:95]: same here

File: vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+            </div>
+        );
+    }
+
+    protected handleSourcecodeLookup = (e: React.MouseEvent<HTMLParagraphElement>): void => this.doHandleSourcecodeLookup(e);
+
+    private doHandleSourcecodeLookup(e: React.MouseEvent<HTMLParagraphElement>) {
+        const { fileLocation, line }: { fileLocation: string, line: string } = JSON.parse(`${e.currentTarget.getAttribute('data-id')}`);
+        console.log('filename: ' + fileLocation + ':' + line);
+        console.log('Source lookup method not implemented');
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/views/properties/vscode-trace-explorer-tooltip-widget.tsx:62]: Ok we can keep that warning. Source code lookup feature needs to be improved and handled in the whole application (consistently) anyways.

**Review by bhufmann — CHANGES_REQUESTED** [review]
Thanks for the contribution. It works with events table and time graph views (tested with Trace Compass trace server).

File: vscode-trace-webviews/src/trace-explorer/properties/vscode-trace-explorer-properties-widget.tsx [hunk 20]
```diff
@@ -0,0 +1,64 @@
+/***************************************************************************************
+ * Copyright (c) 2023 BlackBerry Limited and contributors.
```
Comment (bhufmann) [inline vscode-trace-webviews/src/trace-explorer/properties/vscode-trace-explorer-properties-widget.tsx:2]: `and others` instead of `and contributors`

**Review by bhufmann — CHANGES_REQUESTED** [review]
Small update to the license headers requested.

**Review by bhufmann — APPROVED** [review]
LGTM. Thanks a lot for the contribution.

**Comment by ngondalia** [comment]
@bhufmann  Can you please merge it? I don't have write access to the repo.

**Review by Rodrigoplp-work — APPROVED** [review]

**Pull request merged**